src/crypto/sha256_sse41.cpp

```diff
@@ -6,6 +10,211 @@
+; * Redistributions of source code must retain the above copyright
+;   notice, this list of conditions and the following disclaimer.
+;
+; * Redistributions in binary form must reproduce the above copyright
+;   notice, this list of conditions and the following disclaimer in the
+;   documentation and/or other materials provided with the
+;   distribution.
+;
+; * Neither the name of the Intel Corporation nor the names of its
+;   contributors may be used to endorse or promote products derived from
+;   this software without specific prior written permission.
+;
+; THIS SOFTWARE IS PROVIDED BY INTEL CORPORATION "AS IS" AND ANY
+; EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+; IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+; PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL INTEL CORPORATION OR
+; CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+; EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+; PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+; PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+; LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+; NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+; SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*/
+
+namespace {
+
+uint32_t inline __attribute__((always_inline)) Ror(uint32_t x, int val) { return ((x >> val) | (x << (32 - val))); }
+
+/** Compute one round of SHA256. */
+void inline __attribute__((always_inline)) Round(uint32_t& a, uint32_t& b, uint32_t& c, uint32_t& d, uint32_t& e, uint32_t& f, uint32_t& g, uint32_t& h, uint32_t w)
+{
+    uint32_t t0, t1, t2;
+    t0 = Ror(e, 25 - 11) ^ e;
+    t1 = Ror(a, 22 - 13) ^ a;
+    t2 = (f ^ g) & e;
+    t0 = Ror(t0, 11 - 6) ^ e;
+    t1 = Ror(t1, 13 - 2) ^ a;
+    t0 = Ror(t0, 6);
+    t2 = (t2 ^ g) + t0 + w;
+    t1 = Ror(t1, 2);
+    h += t2;
+    t0 = (a | c) & b;
+    d += h;
+    h += t1;
+    t0 |= (a & c);
+    h += t0;
+}
+
+/** Compute 4 rounds of SHA256, while simultaneously computing the expansion for
+ *  16 rounds later.
+ *
+ *  Input: a,b,c,d,e,f,g,h: The state variables to update with 4 rounds
+ *         x0,x1,x2,x3:     4 128-bit variables containing expansions.
+ *                          If the current round is r, x0,x1,x2,x3 contain the
+ *                          expansions for rounds r..r+15. x0 will be updated
+ *                          to have the expansions for round r+16..r+19.
+ *         W:               The round constants for r..r+3.
+ */
+void inline __attribute__((always_inline)) QuadRoundSched(uint32_t& a, uint32_t& b, uint32_t& c, uint32_t& d, uint32_t& e, uint32_t& f, uint32_t& g, uint32_t& h, __m128i& x0, __m128i x1, __m128i x2, __m128i x3, __m128i w)
+{
+    alignas(__m128i) uint32_t w32[4];
+    __m128i t0, t1, t2, t3, t4;
+
+    w = _mm_add_epi32(w, x0);
+    _mm_store_si128((__m128i*)w32, w);
+
+    Round(a, b, c, d, e, f, g, h, w32[0]);
```

Comment:
is value read from `w32` is guaranteed to be initialised here?

Reply:
Yes, in the line above.